### PR TITLE
feat(diners): add support for Diners Intl

### PR DESCRIPTION
**Pull request by fighou**
#67

- [ ]  @aloisdg
- [ ]  @ytyukhnin

### CODE REVIEW

File: src/Tests/IdentifyTests.cs [hunk 1]
```diff
@@ -192,4 +192,13 @@ public class IdentifyTests
     [TestCase("8600000000500700", ExpectedResult = CardType.UzCard)]
     [TestCase("8600002200510732", ExpectedResult = CardType.UzCard)]
     public CardType ShouldIdentifyAsUzCard(string cardNumber) => Cardidy.Identify(cardNumber).First();
+
+    [TestCase("3634948501962945908", ExpectedResult = CardType.DinersClubInternational)]
+    [TestCase("36148411266187", ExpectedResult = CardType.DinersClubInternational)]
```
Comment (aloisdg) [inline src/Tests/IdentifyTests.cs:197]: Can you add a test for each valid length?

File: src/Dedge.Cardidy/Model/Cards.cs [hunk 2]
```diff
@@ -124,3 +124,13 @@ internal record UzCard : ACard
 {
     public UzCard() : base(CardType.UzCard, 8600, Sixteen) { }
 }
+
+internal record DinersClubInternational : ACard
+{
+    public DinersClubInternational() : base(CardType.DinersClubInternational, 36, From14To19) { }
+}
+
+internal record DinersClubUsAndCanada : ACard
```
Comment (aloisdg) [inline src/Dedge.Cardidy/Model/Cards.cs:133]: Both Diners card are validated with Luhn algorithm. For each, use `ALuhnCard` instead of `ACard`.

**Review by aloisdg — CHANGES_REQUESTED** [review]

File: src/Dedge.Cardidy/Model/Cards.cs (hunk 2 above)
Comment (torendil) [reply src/Dedge.Cardidy/Model/Cards.cs:133]: Didn't catch that, good work

**Review by torendil — APPROVED** [review]
With additional tests, LGTM @aloisdg

File: src/Tests/IdentifyTests.cs [hunk 3]
```diff
@@ -192,4 +192,17 @@ public class IdentifyTests
...
+    [TestCase("361633833419261", ExpectedResult = CardType.DinersClubInternational)]
+    [TestCase("3616338334560012", ExpectedResult = CardType.DinersClubInternational)]
+    [TestCase("36203293000640011", ExpectedResult = CardType.DinersClubInternational)]
+    [TestCase("364540458679150018", ExpectedResult = CardType.DinersClubInternational)]
+    [TestCase("3634948501962945908", ExpectedResult = CardType.DinersClubInternational)]
+    public CardType ShouldIdentifyAsDinersClubInternational(string cardNumber) => Cardidy.Identify(cardNumber).First();
+
+    [TestCase("5495996449417153", ExpectedResult = CardType.DinersClubUsAndCanada)]
+    [TestCase("5422292357387168", ExpectedResult = CardType.DinersClubUsAndCanada)]
+    public CardType ShouldIdentifyAsDinersClubUsAndCanada(string cardNumber) => Cardidy.Identify(cardNumber).First(e => e == CardType.DinersClubUsAndCanada);
```
Comment (aloisdg) [inline src/Tests/IdentifyTests.cs:206]: Why the lambda in the `First` is needed?

Comment (fighou) [reply src/Tests/IdentifyTests.cs:206]: Because Diners can be MasterCard, so Identify returns Diners and MasterCard

Comment (aloisdg) [reply src/Tests/IdentifyTests.cs:206]: True. Alter the test to reveal it.

```
[TestCase("5495996449417153", ExpectedResult =  new[] { CardType.DinersClubUsAndCanada })]
[TestCase("5422292357387168", ExpectedResult = new[] {  CardType.DinersClubUsAndCanada, CardType.MasterCard })]
public IEnumerable<CardType> ShouldIdentifyAsDinersClubUsAndCanada(string cardNumber) => Cardidy.Identify(cardNumber);
```

**Comment by 4gq** [comment]
I would not create the card type DinersClubUsAndCanada as all its cards issued are Mastercard.
In wikipedia page [Diners Club International](https://en.wikipedia.org/wiki/Diners_Club_International), we can see:
`In 2004, Diners Club announced an agreement with Mastercard. Since 2004, all Diners Club cards issued in the United States and Canada (by Citibank until 2009, by BMO since) feature a Mastercard logo and 16-digit account number on the front and can be used wherever Mastercards are accepted. These cards are processed through the Mastercard network. Today's Diners Club in North America, including its sub-brands such as Carte Blanche, is essentially a branded Mastercard.`

And I would rename the DinersClubInternational to DinersClub

**Comment by aloisdg** [comment]
@4gq Some following questions. Does Diners Club american users know that they are basically using a Mastercard? Will they freak out if their Diners Club card are linked to a Mastercard's logo? As is, every card starting with "54" will have both `DinersClubUsAndCanada` and `MasterCard` as result. Which one should be first? We can suppose that most dev using Cardidy will stop at the fist result in the list. What if a MasterCard user's card is identify with a Diners Club's logo?

> rename the DinersClubInternational to DinersClub

When Wikipedia does it, we will follow. We can submit the name change there t quickstart the move.

**Comment by 4gq** [comment]
Hello @aloisdg 

On [google image](https://www.google.com/search?q=mastercard+diners+club&source=lnms&tbm=isch), we can see sample cards with both logos. Note that the cards mainly start with "55".

Also i really doubt that all cards starting with "54" (or "55") are DinersClub. I tried some online card number genarators for Diners Club United States & Canada and all the card numbers were detected as mastercard

**Review by aloisdg — APPROVED** [review]

**Comment by aloisdg** [comment]
Alright let's go with MasterCard then Diners

**Pull request merged**